pkg/util/machines.go

```diff
@@ -179,3 +170,43 @@ func GetMachineMetadata(hostname string, machine infrav1.VSphereMachine, network
 	}
 	return buf.Bytes(), nil
 }
+
+const (
+	// ProviderIDPattern is a regex pattern and is used by ConvertProviderIDToUUID
+	// to convert a providerID into a UUID string.
+	ProviderIDPattern = `(?i)^vsphere://([a-f\d]{8}-[a-f\d]{4}-[a-f\d]{4}-[a-f\d]{4}-[a-f\d]{12})$`
+
+	// UUIDPattern is a regex pattern and is used by ConvertUUIDToProviderID
+	// to convert a UUID into a providerID string.
+	UUIDPattern = `(?i)^[a-f\d]{8}-[a-f\d]{4}-[a-f\d]{4}-[a-f\d]{4}-[a-f\d]{12}$`
+)
+
+// ConvertProviderIDToUUID transforms a provider ID into a UUID string.
+// If providerID is nil, empty, or invalid, then an empty string is returned.
+// A valid providerID should adhere to the format specified by
+// ProviderIDPattern.
+func ConvertProviderIDToUUID(providerID *string) string {
+	if providerID == nil || *providerID == "" {
+		return ""
+	}
+	pattern := regexp.MustCompile(ProviderIDPattern)
+	matches := pattern.FindStringSubmatch(*providerID)
+	if len(matches) < 2 {
+		return ""
+	}
+	return matches[1]
+}
+
+// ConvertUUIDToProviderID transforms a UUID string into a provider ID.
+// If the supplied UUID is empty or invalid then an empty string is returned.
+// A valid UUID should adhere to the format specified by UUIDPattern.
+func ConvertUUIDToProviderID(uuid string) string {
+	if uuid == "" {
+		return ""
+	}
+	pattern := regexp.MustCompile(UUIDPattern)
+	if !pattern.MatchString(uuid) {
+		return ""
+	}
+	return "vsphere://" + uuid
```

Comment:
declare "vsphere://" as const?

Reply:
Hi @figo,

Perhaps, but this is the only place it is used outside of the regex. I suppose declaring it a const and then building the regex with it isn't a bad idea. Thanks!

Reply:
Done and done. Thanks!